src/solvers/refinement/string_refinement.cpp

```diff
@@ -494,35 +602,59 @@ static union_find_replacet string_identifiers_resolution_from_equations(
     "WARNING string_refinement.cpp "
     "string_identifiers_resolution_from_equations:";
 
-  union_find_replacet result;
-  for(const equal_exprt &eq : equations)
+  equation_symbol_mappingt equation_map;
+
+  // Indexes of equations that need to be added to the result
+  std::unordered_set<size_t> required_equations;
+  std::stack<size_t> equations_to_treat;
+
+  for(std::size_t i = 0; i < equations.size(); ++i)
   {
-    if(eq.rhs().type() == string_typet())
-      result.make_union(eq.lhs(), eq.rhs());
-    else if(has_string_subtype(eq.lhs().type()))
+    const equal_exprt &eq = equations[i];
+    if(eq.rhs().id() == ID_function_application)
     {
-      if(eq.rhs().type().id() == ID_struct)
-      {
-        const struct_typet &struct_type = to_struct_type(eq.rhs().type());
-        for(const auto &comp : struct_type.components())
-        {
-          if(comp.type() == string_typet())
-          {
-            const member_exprt lhs_data(eq.lhs(), comp.get_name(), comp.type());
-            const exprt rhs_data = simplify_expr(
-              member_exprt(eq.rhs(), comp.get_name(), comp.type()), ns);
-            result.make_union(lhs_data, rhs_data);
-          }
-        }
-      }
-      else
+      if(required_equations.insert(i).second)
+        equations_to_treat.push(i);
+
+      std::vector<exprt> rhs_strings = extract_strings(eq.rhs());
+      for(const auto expr : rhs_strings)
+        equation_map.add(i, expr);
+    }
+    else if(eq.lhs().type().id() != ID_pointer &&
+       has_string_subtype(eq.lhs().type()))
+    {
+      std::vector<exprt> lhs_strings = extract_strings_from_lhs(eq.lhs());
+
+      for(const auto expr : lhs_strings)
+        equation_map.add(i, expr);
+
+      if(lhs_strings.empty())
       {
-        stream << log_message << "non struct with string subexpr "
-               << from_expr(ns, "", eq.rhs()) << "\n  * of type "
-               << from_type(ns, "", eq.rhs().type()) << eom;
+        stream << log_message << "non struct with string subtype "
+               << from_expr(ns, "", eq.lhs()) << "\n  * of type "
+               << from_type(ns, "", eq.lhs().type()) << eom;
       }
+
+      for(const exprt &expr : extract_strings(eq.rhs()))
+        equation_map.add(i, expr);
     }
   }
+
+  // transitively add all equation which depends on the equations to treat
+  while(!equations_to_treat.empty())
+  {
+    const std::size_t i = equations_to_treat.top();
+    equations_to_treat.pop();
+    for(const exprt &string : equation_map.find_expressions(i))
```

Comment:
Nit since these for loops spread over multiple lines please add braces for clarity